airbyte-integrations/connectors/source-shopify/source_shopify/shopify_graphql/bulk/job.py

```diff
@@ -52,12 +56,23 @@ class ShopifyBulkManager:
     # attempt counter
     concurrent_attempt: int = field(init=False, default=0)
 
-    # default logger
-    logger: AirbyteLogger = field(init=False, default=logging.getLogger("airbyte"))
-
     # currents: job_id, job_state
     job_id: Optional[str] = field(init=False, default=None)
     job_state: ShopifyBulkStatus = field(init=False, default=None)
+    # Each job ideally should be executed within the specified time (in sec),
+    # to maximize the performance for multi-connection syncs and control the bulk job size within +- 7 mins (420 sec),
+    # Ideally the source will balance on it's own rate, based on the time taken to return the data for the slice.
+    job_elapsed_time_threshold_sec: float = field(init=False, default=420.0)
```

Comment:
Overall, it seems like we have a lot of stateful variables which makes is hard to create a mental model. Would it be possible to delegate the slice_range to the job manager? Something like:

```
class IncrementalShopifyGraphQlBulkStream(IncrementalShopifyStream):
    @stream_state_cache.cache_stream_state
    def stream_slices(self, stream_state: Optional[Mapping[str, Any]] = None, **kwargs) -> Iterable[Optional[Mapping[str, Any]]]:
        <...>
            while start < end:
                slice_end = start.add(days=self.slice_interval)
            <...>
            
<...>
    @property
    def slice_interval(self) -> timedelta:
        return days=self.job_manager.slice_interval
```

... and:

```
class ShopifyBulkManager:
    <...>
    def job_check(self, created_job_response: requests.Response) -> Optional[str]:
        <...>
        finally:
            current_job_elapsed_time = round((time() - job_started), 3)
            if current_job_elapsed_time > self.job_elapsed_time_threshold_sec:
                self.slice_interval_in_days /= self._clamp(self.slice_internal_in_days/2)
            elif current_job_elapsed_time < 1 or current_job_elapsed_time < self.job_last_elapsed_time_sec:
                self.slice_size_expand_factor = self._calculate_expand_factor()
                self.slice_interval_in_days = self._clamp(self.slice_interval_in_days + self.slice_size_expand_factor)

            self.job_last_elapsed_time_sec = current_job_elapsed_time
            self.logger.info(f"The BULK Job: `{self.job_id}` time elapsed: {self.current_job_elapsed_time} sec.")
            # reset the state for COMPLETED job
            self.__reset_state()

    def _clamp(self, value, lower=None, upper=None):
        if lower is None:
            lower = self.slice_interval_in_days_min
        if upper is None:
            upper = self.slice_interval_in_days_max
        return max(lower, min(value, upper))
```

This way, I think we can remove in ShopifyBulkManager:
* current_job_elapsed_time
* job_should_expand_slice_size
* job_should_reduce_slice_size
* slice_size_reduce_factor

... which means four less variables that the reader needs to understand and keep track of.

Note: I assume there is a reason for all this stateful setup but I can't see it so please share this with me so that we can make it explicit in the code as well